### ISSUE TITLE
Type-based termination

### ISSUE THREAD

**Issue by GoogleCodeExporter** [agda/agda#695]
Consider the following program:
```agda
module Termination where

data Sigma (A : Set)(B : A -> Set) : Set where
  _,_ : (x : A) -> B x -> Sigma A B

record Top : Set where

_o_ : {A B : Set}{C : Set1} ->
      (f : B -> C) -> (g : A -> B) -> (A -> C)
f o g = \ x -> f (g x)


mutual
  data U  : Set where
    top : U
    sig : (X : U) -> (T X -> U) -> U


  T : U -> Set
  T top = Top
--  T (sig a b) = Sigma (T a) (\ x -> T (b x))
  T (sig a b) = Sigma (T a) (T o b)
```
Using the latest darcs version of Agda, the termination checker complains that `T` is possibly non-terminating. Changing the line for `T (sig a b)` to the commented out one (with `T o b` expanded) makes the program pass, but it would be nice if also the original program would be accepted.

Adding `{-# OPTIONS --termination-depth=100 #-}` does not help.

Original issue reported on code.google.com by `fredrik....@gmail.com` on 17 Sep 2012 at 3:02

**Comment by nad** [comment]
>> Are there any plans to automatically infer such annotations?

> Yes, we have plans, but it might still take a while to implement.

Any concrete ideas?

**Comment by UlfNorell** [comment]
This example is fixed by 2a349528 (compile-time auto inlining).

**Comment by UlfNorell** [comment]
I don't think we have any plans at the moment to make everything sized-typed. If or when these materialise we can open a new issue.
